cli/util/draw_thread.rs

```diff
@@ -0,0 +1,237 @@
+/// static place on the console.
+pub trait DrawThreadRenderer: Send + Sync + std::fmt::Debug {
+  fn render(&self, data: &ConsoleSize) -> String;
+}
+
+/// Draw thread guard. Keep this alive for the duration
+/// that you wish the entry to be drawn for. Once it is
+/// dropped, then the entry will be removed from the draw
+/// thread.
+#[derive(Debug)]
+pub struct DrawThreadGuard(u16);
+
+impl Drop for DrawThreadGuard {
+  fn drop(&mut self) {
+    DrawThread::finish_entry(self.0)
+  }
+}
+
+#[derive(Debug, Clone)]
+struct InternalEntry {
+  priority: u8,
+  id: u16,
+  renderer: Arc<dyn DrawThreadRenderer>,
+}
+
+#[derive(Debug)]
+struct InternalState {
+  // this ensures only one actual draw thread is running
+  drawer_id: usize,
+  hide: bool,
+  has_draw_thread: bool,
+  next_entry_id: u16,
+  entries: Vec<InternalEntry>,
+  static_text: ConsoleStaticText,
+}
+
+impl InternalState {
+  pub fn should_exit_draw_thread(&self, drawer_id: usize) -> bool {
+    self.drawer_id != drawer_id || self.entries.is_empty()
+  }
+}
+
+static INTERNAL_STATE: Lazy<Arc<Mutex<InternalState>>> = Lazy::new(|| {
+  Arc::new(Mutex::new(InternalState {
+    drawer_id: 0,
+    hide: false,
+    has_draw_thread: false,
+    entries: Vec::new(),
+    next_entry_id: 0,
+    static_text: ConsoleStaticText::new(|| {
+      let size = console_size().unwrap();
+      console_static_text::ConsoleSize {
+        cols: Some(size.cols as u16),
+        rows: Some(size.rows as u16),
+      }
+    }),
+  }))
+});
+
+/// The draw thread is responsible for rendering multiple active
+/// `DrawThreadRenderer`s to stderr. It is global because the
+/// concept of stderr in the process is also a global concept.
+#[derive(Clone, Debug)]
+pub struct DrawThread;
+
+impl DrawThread {
+  /// Is using a draw thread supported.
+  pub fn is_supported() -> bool {
+    atty::is(atty::Stream::Stderr)
+      && log::log_enabled!(log::Level::Info)
+      && console_size()
+        .map(|s| s.cols > 0 && s.rows > 0)
+        .unwrap_or(false)
+  }
+
+  /// Adds a renderer to the draw thread with a given priority.
+  /// Renderers are sorted by priority with higher priority
+  /// entries appearing at the bottom of the screen.
+  pub fn add_entry(
+    priority: u8,
+    renderer: Arc<dyn DrawThreadRenderer>,
+  ) -> DrawThreadGuard {
+    let internal_state = &*INTERNAL_STATE;
+    let mut internal_state = internal_state.lock();
+    let id = internal_state.next_entry_id;
+    internal_state.entries.push(InternalEntry {
+      id,
+      priority,
+      renderer,
+    });
+    internal_state
+      .entries
+      .sort_by(|a, b| a.priority.cmp(&b.priority));
+
+    if internal_state.next_entry_id == u16::MAX {
+      internal_state.next_entry_id = 0;
+    } else {
+      internal_state.next_entry_id += 1;
+    }
+
+    Self::maybe_start_draw_thread(&mut internal_state);
+
+    DrawThreadGuard(id)
+  }
+
+  /// Hides the draw thread.
+  #[allow(dead_code)]
+  pub fn hide() {
```

Comment:
Hide and show functionality is here. I did not test these, but it should be trivial to fix any mistakes if there are any once using this in the permission prompt pr.

Reply:
How would we call it in the prompt callback?

Reply:
`DrawThread::hide()`

Reply:
Since the refactor PR of permission prompt has landed and these callbacks are now available maybe we should hook them up in this PR? That way we won't have to land with the `#[allow(dead_code)]` attribute

Reply:
Let's land this, then I'll open a new PR just for that.